src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs

```diff
@@ -545,26 +545,34 @@ public ImmutableArray<T> InsertRange(int index, IEnumerable<T> items)
             Requires.NotNull(items, "items");
 
             if (self.Length == 0)
-            {
                 return ImmutableArray.CreateRange(items);
-            }
-
-            int count = ImmutableExtensions.GetCount(ref items);
-            if (count == 0)
+            
+            if (items is ICollection<T>)
             {
-                return self;
+                ICollection<T> other = (ICollection<T>)items;
+                
+                if (other.Count == 0)
+                    return self;
+                
+                T[] dest = new T[self.Length + other.Count];
+                Array.Copy(self.array, 0, dest, 0, index);
+                other.CopyTo(dest, index);
```

Comment:
:exclamation: There is potential ability for an implementation of `CopyTo` to manipulate elements of the array which it shouldn't. You can solve this by moving the call to `other.CopyTo` above the two calls to `Array.Copy`, ensuring that even if the call to `CopyTo` results in out-of-bounds array writes, the elements before and after the inserted range are fully protected against modification.